enhancements/single-node-production-deployment-approach.md

```diff
@@ -0,0 +1,484 @@
+* This enhancement does not presume that (more) operators are modified
+  to "render" manifests while the cluster is being built so that the
+  operators do not have to be run in the cluster.
+* This enhancement does not attempt to describe a way to "pre-build"
+  deployment images, either generically or customized for a user.
+* This enhancement does not address the removal of the bootstrap VM,
+  although single-node clusters would benefit from that work, which
+  will be described in a separate enhancement.
+
+## Proposal
+
+After the [capabilities API is
+introduced](https://github.com/openshift/enhancements/pull/555), all
+teams developing OpenShift components will need to consider how their
+components should be configured when deployed and used when the
+high-availability capability is disabled.
+
+In the telco RAN use case, high-availability is typically achieved by
+having multiple sites provide coverage to overlapping geographic
+areas. Therefore, use of cluster-based high-availability features will
+be limited (for example, by running a single API service).
+
+Failures in edge deployments are frequently resolved by re-imaging or
+physically replacing the entire host. Combining this fact with the
+previous observation about the approach to providing highly-available
+services lets us draw the conclusion that in-place upgrades do not
+need to be supported by the first iteration of this configuration.
+
+In a multi-node cluster, operators may safely trigger rolling reboots
+or restarts of the control plane services without taking the cluster
+offline. In a single-node deployment, any such restart would take some
+or all of the services offline. The deployment or upgrade of a cluster
+may include several host reboots, clustered together in time, until
+the operators stabilize. Users should wait for this period of
+instability before installing their workloads.
+
+The single-node configuration of OpenShift will be sufficiently
+generic to cater to a variety of edge computing use cases. As such,
+OpenShift's usual cluster configuration mechanisms will be favored
+where there is a likelihood that there will be edge computing use case
+with differing requirements.  For example, there is no expectation
+that single-node deployments will use the real-time kernel by
+default -- this will continue to be a `MachineConfig` choice as per
+[enhancements/support-for-realtime-kernel](https://github.com/openshift/enhancements/blob/master/enhancements/support-for-realtime-kernel.md). *See open questions*
+
+### User Stories
+
+#### As a user, I can deploy OpenShift in a supported single-node configuration
+
+A user will be able to run the OpenShift installer to create a single-node
+deployment, with some limitations (see non-goals above). The user
+will not require special support exceptions to receive technical assistance
+for the features supported by the configuration.
+
+### Implementation Details/Notes/Constraints
```

Comment:
Do we need to make any changes to `oc`, CVO, or the console to make it clear that upgrades are not allowed? Or is it sufficient for the dev preview to just document this limitation?

Reply:
I don't know what the standard policy for that is. Do we generally support upgrades from deployments using dev-preview features?

Reply:
@rhamilto for console cluster settings upgrade visualization expertise

Reply:
The console looks to the CVO for upgrade information.  So if upgrading is not allowed, the CVO should control that.

Reply:
right, I'm not suggesting we control it but it might be useful to note what can/can't be displayed today.  We have the notification that would show an alert if the CVO service monitor fires it's [ClusterNotUpgradeable alert](https://github.com/openshift/cluster-version-operator/blob/1e51a0e4750ca110d4659f33bce210a3de6844b9/install/0000_90_cluster-version-operator_02_servicemonitor.yaml#L61-L70), right?  Anything on the cluster settings screens that you revamped?

Reply:
> Anything on the cluster settings screens that you revamped?

Nope.  You only see the upgrade UI elements if you can upgrade and an upgrade is available.  We don't explicitly tell you you can't upgrade there.

Reply:
Does the CVO know that a cluster is using a dev-preview feature and prevent upgrades? Or do we just tell users that we don't support them if they do it? If we actively block it, can someone point me to docs for how that works so we can make sure we have something in place to set it up?

Reply:
> Does the CVO know that a cluster is using a dev-preview feature and prevent upgrades? Or do we just tell users that we don't support them if they do it? If we actively block it, can someone point me to docs for how that works so we can make sure we have something in place to set it up?

That I do not know.  @wking, do you?

Reply:
I've add an open question about how to deal with attempts to upgrade.